Blog post about Telemetry database and Metabase (#177)
readthedocs/blog: knowing-more-about-ourselves.rst

We decided to `start working on a project <https://github.com/readthedocs/readthedocs.org/pull/8124>`_
to collect relevant data to answer a large number questions about how people use our service.

What data are we collecting?
----------------------------

As you may already know, Read the Docs has a strict `privacy policy <https://docs.readthedocs.io/en/stable/privacy-policy.html>`_
which we follow to ensure we are protecting the privacy of our users, authors and readers.
It's still worth mentioning that most of the data we collect is publicly available,
or is generated by our own service.

The data generated by Read the Docs which we already had access to was mostly high-level about our database objects.
A few examples would be:

* project configuration (e.g. PR build enabled, default branch)
* build metadata (e.g. number of builds, length of builds)
* platform information (redirects, webhooks, search queries, pageviews)

With our new data collections project,
we were able to collect a lot of new, interesting data.
This new data allows us to answer a whole new set of questions,
and is mainly related to the build process
The new information that we're collecting includes:

> * Python dependencies installed during a documentation build
> * Apt packages installed prior to building docs
> * The entire contents of the `Read the Docs configuration file <https://docs.readthedocs.io/en/stable/config-file/index.html>`_

Author: Now we are using a bullet list, we can add more things here if we want:

```suggestion
* OS and Python version used to build the documentation
* Python and Conda dependencies installed during a documentation build
* Apt packages installed prior to building docs
* The entire contents of the `Read the Docs configuration file <https://docs.readthedocs.io/en/stable/config-file/index.html>`_
```

I didn't include them original because it was just a paragraph and I they weren't too relevant. I'm fine if we skip them, tho... They aren't too relevant right at the moment.

We connected our PostgreSQL replica database to it and started plotting this data to get some insights.
We found awesome things we want to start sharing with all the documentation community,
since we think it will useful not only for ourselves but also for package maintainers.

Sphinx extensions used in the last 15 days
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. image:: /img/sphinx-extensions.png
   :align: center
   :width: 100%

.. note::

   We excluded ``sphinx-rtd-theme`` and ``sphinxcontrib-websupport`` because they are too high and break the relation with the others in the graph.
   Also, only extensions used by more than 10 projects are showed.

Sphinx adoption over time
~~~~~~~~~~~~~~~~~~~~~~~~~

.. image:: /img/sphinx-adoption.png
   :align: center
   :width: 100%

Projects using MyST to write Markdown in Sphinx
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. image:: /img/myst-parser.png
   :align: center
   :width: 100%

VCS types used by projects with successful builds in last year
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. image:: /img/project-vcs-types.png
   :align: center
   :width: 100%

> Docker image used in the last 15 days

Reviewer: What is `default` here? Also, is `ubuntu-latest` a real option, or an error?

Author: `default` means the user did not specify anything in their config file. IIRC, on config file v2, this is `latest`.

`ubuntu-latest` is an invalid image and the build should have failed. I'm not sure why it's listed here. I didn't notice it when taking the screenshot. That's wrong and it shouldn't be there. I will take a look at the query and fix it. Thanks!

Author: Hrm... I checked the query. The user used `build.os: ubuntu-latest`, and the config was translated finally to `build.image: readthedocs/build:latest`. That's why it succeed. However, this is an error and we should raise a build error or similar.

I will remove it from this query for now because I don't want to communicate `ubuntu-latest` on the blog post since it's an invalid option.

Author: This is done. I updated the plot with more accurate data.

.. post:: September 13, 2022
   :tags: metabase, data, grant
   :author: Manuel
   :location: BCN

.. meta::
   :description lang=en:

>       Collect and analyze project and build data to make better decision about our product
>       with a strong focus on the user experience.

Reviewer 2: ```suggestion
      Collect and analyze project and build data to make better decisions about our product
      with a strong focus on the user experience.
```

Last year, with the growth of our product and the team,
plus the `CZI grant we received <https://blog.readthedocs.com/czi-grant-announcement/>`_,
we started asking ourselves some questions that we couldn't answer with the data we had.
We decided to `start working on a project <https://github.com/readthedocs/readthedocs.org/pull/8124>`_
to collect relevant data to answer a large number of questions about how people use our service.

What data are we collecting?
----------------------------

As you may already know, Read the Docs has a strict `privacy policy <https://docs.readthedocs.io/en/stable/privacy-policy.html>`_
which we follow to ensure we are protecting the privacy of our users, authors and readers.
It's still worth mentioning that most of the data we collect is publicly available,
or is generated by our own service.

The data generated by Read the Docs which we already had access to was mostly high-level about our database objects.
A few examples would be:

* project configuration (e.g. PR build enabled, default branch)
* build metadata (e.g. number of builds, length of builds)
* platform information (redirects, webhooks, search queries, pageviews)

> With our new data collections project,

Reviewer 2: ```suggestion
With our new data collection project,
```

Note that versions marked with a red rectangle are the most relevant ones for this analysis,
and are highlighted just to improve readability.

There is something that's pretty prominent on this plot as well.
The old version 1.8.6 has around 1000 projects, which could look suspicious at first sight.
However, this is because Read the Docs still defaults to 1.8.6 on projects created before Oct 20, 2020.
This is to keep compatibility with old projects and avoid breaking their builds without notification.

Projects using MyST to write Markdown in Sphinx
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. image:: /img/myst-parser.png
   :align: center
   :width: 100%

This is a similar plot than the previous one but for `MyST parser <https://myst-parser.readthedocs.io/>`_.
We are interested in knowing how many projects are writing Markdown using Sphinx.
Many people don't know it's possible to write Markdown while keep using Sphinx and all the amazing features it has.
Interestingly, the number of projects using MyST parser, currently, is higher than the number of projects using MkDocs.
If you are reading this and you didn't know about MyST parser,

> we strongly recommend you to give it a try!

Reviewer 2: This is a good place for a link to a guide on using MyST

Author: I put the link at the beginning of this paragraph already. Should I link this last sentence to the "Get started" guide? https://myst-parser.readthedocs.io/en/latest/intro.html

      Collect and analyze project and build data to make better decision about our product
      with a strong focus on the user experience.

Knowing more about how people use our service
=============================================

Read the Docs generates a lot of data.
We are the largest documentation platform out there,

> with hundreds of thousands of projects using our product every day to host their documentation.

Reviewer: This still reads weirdly, but I couldn't think of an obvious way to state it better. Maybe:

`with 250,000+ projects`

Author: I don't have a preference here. Both of them are readable to my level of English 😄

   :width: 100%

> Thanks to this plot we were able to find some anomalies in the last weeks.
> We noticed that there were a few projects consuming 3x build time that their successor.
> This was due to the development workflow they follow (multiple pushes with small commits),
> making our platform to trigger multiple builds where each of them took more than 15 minutes to complete.
> This generated a bad UX, since they had to wait for *all the old builds to finish*
> before being able to see the results from the latest builds --  which was the only valid one.
> 
> After noticing this happening pretty frequently to these projects,
> we prioritized the work on `Cancel old builds <https://github.com/readthedocs/readthedocs.org/issues/8961>`_,
> reducing the computing time for our servers and the waiting time for our users: Win-Win!

Reviewer: ```suggestion
Thanks to this plot we were able to find some anomalies in our builds.
We noticed that there were a few projects consuming 3x build time between different builds.
This was due to the development workflow they follow (multiple pushes with small commits),
making our platform to trigger multiple builds where each of them took more than 15 minutes to complete.
This caused a bad UX, since they had to wait for *all the old builds to finish*
before being able to see the results from the latest builds --  which was the only valid one.

After noticing this happening pretty frequently to these projects,
we prioritized the work on `Cancel old builds <https://github.com/readthedocs/readthedocs.org/issues/8961>`_,
reducing the computing time for our servers and the waiting time for our users: Win-Win!
```
